tests/PactNet.Native.Tests/NativeRequestBuilderTests.cs

```diff
@@ -94,15 +94,41 @@ public void WithQuery_RepeatedQuery_SetsIndex()
         }
 
         [Fact]
-        public void WithHeader_WhenCalled_AddsHeaderParam()
+        public void WithHeader_Matcher_WhenCalled_AddsSerialisedHeaderParam()
+        {
+            this.builder.WithHeader("name", PactNet.Matchers.Match.Regex("header", "^header$"));
+
+            var expectedValue = "{\"pact:matcher:type\":\"regex\",\"value\":\"header\",\"regex\":\"^header$\"}";
+
+            this.mockServer.Verify(s => s.WithRequestHeader(this.handle, "name", expectedValue, 0));
+        }
+
+        [Fact]
+        public void WithHeader_Matcher_RepeatedHeader_SetsIndex()
+        {
+            this.builder.WithHeader("name", PactNet.Matchers.Match.Regex("value1", "^value1$"));
+            this.builder.WithHeader("name", PactNet.Matchers.Match.Type("value2"));
+            this.builder.WithHeader("other", PactNet.Matchers.Match.Regex("value", "^value$"));
+
+            var expectedValue1 = "{\"pact:matcher:type\":\"regex\",\"value\":\"value1\",\"regex\":\"^value1$\"}";
+            var expectedValue2 = "{\"pact:matcher:type\":\"type\",\"value\":\"value2\"}";
+            var expectedValue = "{\"pact:matcher:type\":\"regex\",\"value\":\"value\",\"regex\":\"^value$\"}";
+
+            this.mockServer.Verify(s => s.WithRequestHeader(this.handle, "name", expectedValue1, 0));
+            this.mockServer.Verify(s => s.WithRequestHeader(this.handle, "name", expectedValue2, 1));
+            this.mockServer.Verify(s => s.WithRequestHeader(this.handle, "other", expectedValue, 0));
+        }
+
+        [Fact]
+        public void WithHeader_String_WhenCalled_AddsHeaderParam()
         {
             this.builder.WithHeader("name", "value");
 
             this.mockServer.Verify(s => s.WithRequestHeader(this.handle, "name", "value", 0));
         }
 
         [Fact]
-        public void WithHeader_RepeatedHeader_SetsIndex()
+        public void WithHeader_String_RepeatedHeader_SetsIndex()
```

Comment:
The naming convention for the tests is Method_PreCondition_PostCondition so I'd call this WithHeader_RepeatedStringHeader_SetsIndex.

We'll need additional tests for RepeatedMatcherHeader as well, and maybe one that does both a string and a matcher together.

Edit: sorry just seen the matcher tests below, so just the naming thing